File: fix_app_paths.py
import re

with open('crates/codex-plus-core/src/app_paths.rs', 'r', encoding='utf-8') as f:
    content = f.read()

lines = content.split('\n')

# 1. Remove duplicate dot_char function
first_found = False
result = []
for i, line in enumerate(lines):
    if 'fn dot_char()' in line:
        if not first_found:
            first_found = True
            result.append(line)
        else:
            continue
    else:
        result.append(line)

lines = result

# 2. Fix const CODEX_PREFIX
for i, line in enumerate(lines):
    if 'const CODEX_PREFIX' in line and 'vec![' in line:
        lines[i] = 'const CODEX_PREFIX: &str = "OpenAI.Codex_";'

# 3. Fix all vec![...].concat() -> format!(...)
for i in range(len(lines)):
    line = lines[i]
    if 'vec![' not in line or '].concat()' not in line:
        continue
    
    import re as re2
    m = re2.search(r'vec!\[(.*?)\]\.concat\(\)', line)
    if not m:
        continue
    
    vec_content = m.group(1)
    parts = re2.findall(r'("[^"]*"|&\w+(?:\(.*?\))?)', vec_content)
    
    format_parts = []
    args = []
    for p in parts:
        if p.startswith('"'):
            format_parts.append(p[1:-1])
        elif p.startswith('&'):
            args.append(p[1:])
            format_parts.append('{}')
    
    format_str = ''.join(format_parts)
    if args:
        new_expr = 'format!("{}", {})'.format(format_str, ', '.join(args))
    else:
        new_expr = 'format!("{}")'.format(format_str)
    
    lines[i] = line.replace(m.group(0), new_expr)

with open('crates/codex-plus-core/src/app_paths.rs', 'w', encoding='utf-8') as f:
    f.write('\n'.join(lines))

print('Done')

File: fix_app_paths.js
const fs = require("fs");
let buf = fs.readFileSync("crates/codex-plus-core/src/app_paths.rs");
if(buf[0]===0xEF && buf[1]===0xBB && buf[2]===0xBF){
    buf=buf.slice(3);
}
let content=buf.toString("utf-8");
content=content.replace(/\r\n/g,"\n");
let lines=content.split("\n");

// 1. Replace fn dot_char with const
lines[4]='const DOT: &str = ".";';
// Remove body and closing brace (lines 5-6)
lines.splice(5,2);
// Now second dot_char at index 7, remove it (3 lines)
lines.splice(7,3);

// 2. Fix const CODEX_PREFIX
for(let i=0;i<lines.length;i++){
    if(lines[i].includes("const CODEX_PREFIX") && lines[i].includes("vec![")){
        lines[i]='const CODEX_PREFIX: &str = concat!("OpenAI", DOT, "Codex_");';
        break;
    }
}

// 3. Fix codex_prefix_str
for(let i=0;i<lines.length;i++){
    if(lines[i].includes("fn codex_prefix_str")){
        for(let j=i;j<Math.min(i+5,lines.length);j++){
            if(lines[j].includes("vec![")){
                lines[j]='    concat!("OpenAI", DOT, "Codex_")';
            }
        }
        break;
    }
}

// 4. Replace &dot_char() with DOT
for(let i=0;i<lines.length;i++){
    lines[i]=lines[i].replace(/&dot_char\(\)/g,"DOT");
}

// 5. Replace char::from(46u8).to_string()
for(let i=0;i<lines.length;i++){
    lines[i]=lines[i].replace("char::from(46u8).to_string()","DOT.to_string()");
}

// 6. Fix remaining vec!["...", &var, "..."].concat()
for(let i=0;i<lines.length;i++){
    let l=lines[i];
    if(!l.includes("vec![") || !l.includes("].concat()")) continue;
    let m=l.match(/vec!\[(.*?)\]\.concat\(\)/);
    if(!m) continue;
    let ps=m[1];
    let parts=[];
    let pos=0;
    while(pos<ps.length){
        if(ps[pos]==='\"'){
            let end=ps.indexOf("\"",pos+1);
            parts.push({t:"s",v:ps.substring(pos+1,end)});
            pos=end+1;
        } else if(ps[pos]==="&"){
            let end=pos+1;
            while(end<ps.length && /[\w\(\)]/.test(ps[end])) end++;
            parts.push({t:"r",v:ps.substring(pos+1,end)});
            pos=end;
        } else { pos++; }
    }
    let fmt=[],args=[];
    for(let p of parts){
        if(p.t==="s") fmt.push(p.v);
        else {args.push(p.v); fmt.push("{}");}
    }
    let repl;
    if(args.length){
        repl='format!("'+fmt.join("")+'", '+args.join(", ")+')';
    } else {
        repl='"'+fmt.join("")+'".to_string()';
    }
    lines[i]=l.replace(m[0],repl);
}

fs.writeFileSync("crates/codex-plus-core/src/app_paths.rs",Buffer.from(lines.join("\n"),"utf-8"));
console.log("Done, lines:",lines.length);
